src/representations/representation_manager_spherical_covariants.hh

```diff
@@ -0,0 +1,500 @@
+                  for (size_t l2{0}; l2 < this->max_angular + 1; l2++) {
+                    if (l1 < static_cast<size_t>(std::abs<int>(l2 - l3)) ||
+                        l1 > l2 + l3) {
+                      continue;
+                    }
+                    if (this->inversion_symmetry == true) {
+                      if ((l1 + l2 + l3) % 2 == 1) {
+                        continue;
+                      }
+                    }
+                    for (size_t m3{0}; m3 < 2 * l3 + 1; m3++) {
+                      int m3s{static_cast<int>(m3 - l3)};
+                      for (size_t m1{0}; m1 < 2 * l1 + 1; m1++) {
+                        int m1s{static_cast<int>(m1 - l1)};
+                        size_t lm1{pow(l1, 2_z) + m1};
+                        for (size_t m2{0}; m2 < 2 * l2 + 1; m2++) {
+                          int m2s{static_cast<int>(m2 - l2)};
+                          if (m1s + m2s + m3s != 0 && m1s + m2s - m3s != 0) {
+                            continue;
+                          }
+                          size_t lm2{pow(l2, 2_z) + m2};
+                          complex coef1c, coef2c;
+                          double w3j = w3js[count];
+                          // usual formulae for converting from real to complex
+                          if (m1s > 0) {
+                            coef1c = pow(-1.0, m1s) *
+                                     complex(coef1(n1, lm1),
+                                             coef1(n1, lm1 - 2 * m1s));
+                          } else if (m1s == 0) {
+                            coef1c =
+                                complex(coef1(n1, lm1), 0.0) * std::sqrt(2.0);
+                          } else if (m1s < 0) {
+                            coef1c = complex(coef1(n1, lm1 - 2 * m1s),
+                                             -coef1(n1, lm1));
+                          }
+                          if (m2s > 0) {
+                            coef2c = pow(-1.0, m2s) *
+                                     complex(coef2(n2, lm2),
+                                             coef2(n2, lm2 - 2 * m2s));
+                          } else if (m2s == 0) {
+                            coef2c =
+                                complex(coef2(n2, lm2), 0.0) * std::sqrt(2.0);
+                          } else if (m2s < 0) {
+                            coef2c = complex{coef2(n2, lm2 - 2 * m2s),
+                                             -coef2(n2, lm2)};
+                          }
+                          coef1c /= std::sqrt(2.0);
+                          coef2c /= std::sqrt(2.0);
+                          // combine the coefficients with Wigner 3j symbols
+                          complex ci{0.0, 1.0};
+                          if ((l1 + l2 + l3) % 2 == 0) {
+                            if (m3s < 0) {
+                              soap_vector_by_type(nn, l0) +=
+                                  w3j * (ci * coef1c * coef2c).real();
```

Comment:
Maybe we can try this later, but as I said above, there must be a way to write this in terms of purely real arithmetic.